src/allmydata/storage/http_server.py

```diff
@@ -501,3 +516,52 @@ def advise_corrupt_share(self, request, authorization, storage_index, share_numb
         info = self._read_encoded(request)
         bucket.advise_corrupt_share(info["reason"].encode("utf-8"))
         return b""
+
+
+def listen_tls(
+    reactor,
+    server: HTTPServer,
+    hostname: str,
+    port: int,
+    private_key_path: Path,
+    cert_path: Path,
+    interface: Optional[str],
+) -> Deferred[Tuple[DecodedURL, IListeningPort]]:
+    """
+    Start a HTTPS storage server on the given port, return the fURL and the
+    listening port.
+
+    The hostname is the external IP or hostname clients will connect to; it
+    does not modify what interfaces the server listens on.  To set the
+    listening interface, use the ``interface`` argument.
+
+    Port can be 0 to choose a random port.
+    """
+    endpoint_string = "ssl:privateKey={}:certKey={}:port={}".format(
+        quoteStringArgument(str(private_key_path)),
+        quoteStringArgument(str(cert_path)),
+        port,
+    )
+    if interface is not None:
+        endpoint_string += ":interface={}".format(quoteStringArgument(interface))
+    endpoint = serverFromString(reactor, endpoint_string)
```

Comment:
Dang. Too bad there still is no decent Python API for constructing a TLS server endpoint. :/

Reply:
besides `SSL4ServerEndpoint(...)`...?
(maybe this provides more inspiration? https://github.com/crossbario/crossbar/blob/master/crossbar/common/twisted/endpoint.py#L107 )

Reply:
Sorry, I lost the plot here: https://twistedmatrix.com/trac/ticket/9885

Reply:
> besides SSL4ServerEndpoint(...)...?

Right, besides that.  If that were the good Python API I imagine then I suppose it would be easier to use it than to build up a string and then parse it with `serverFromString`.

Reply:
> Sorry, I lost the plot here: https://twistedmatrix.com/trac/ticket/9885

Me too.  I couldn't find the mailing list post that ticket used to reference.  Since `wrapClientTLS` is an API that makes TLS composable with other endpoints, maybe twisted:9885 is about a `wrapServerTLS` that works similarly?

That would be nice, certainly, and would let `listen_tcp` have a signature that accepts another endpoint instead of a hostname / port number (or would at least remove one barrier to doing this - the other being that the code still needs a way to describe whatever endpoint it is listening on, which is not possible for an arbitrary endpoint, at least if you actually adhere to the interface).

Reply:
Ah, good point. From memory the gist is https://github.com/glyph/txsni/blob/master/txsni/tlsendpoint.py + write a strports plugin.

Reply:
I dug it up, linked and quoted on the ticket now.

Reply:
I'm just going to live with this for now, since it works.